File: frontend/src/utils/validators.js
// Validation utilities

import { VALID_VEHICLE_CLASSES } from "../data/data"

export const validators = {
  email: (email) => {
    const regex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
    return regex.test(email)
  },

  password: (password) => {
    if (password.length < 8) return 'Password must be at least 8 characters'
    if (!/[A-Z]/.test(password)) return 'Password must contain uppercase letter'
    if (!/[0-9]/.test(password)) return 'Password must contain digit'
    if (!/[!@#$%^&*()-_=+[\]{}|;:',.<>?/]/.test(password))
      return 'Password must contain special character'
    return null
  },

  fullName: (name) => {
    if (name.trim().length < 2) return 'Full name must be at least 2 characters'
    return null
  },

  engineSize: (size) => {
    if (size <= 0 || size > 10) return 'Engine size must be between 0.1 and 10 L'
    return null
  },

  cylinders: (cyl) => {
    if (cyl < 1 || cyl > 16) return 'Cylinders must be between 1 and 16'
    return null
  },

  fuelConsumption: (consumption, field = 'Fuel consumption') => {
    if (consumption <= 0 || consumption > 100)
      return `${field} must be between 0.1 and 100 L/100km`
    return null
  },

  vehicleClass: (vClass) => {
    const valid = VALID_VEHICLE_CLASSES;
    if (!valid.includes(vClass.toUpperCase()))
      return `Vehicle class must be one of: ${valid.join(', ')}`
    return null
  },

  fuelType: (fuelType) => {
    const valid = ['X', 'Z', 'D', 'E', 'N'];
    if (!valid.includes(fuelType.toUpperCase()))
      return `Fuel type must be one of: ${valid.join(', ')}`
    return null
  },

  textField: (value, fieldName = 'Field', minLength = 2) => {
    if (value.trim().length < minLength)
      return `${fieldName} must be at least ${minLength} characters`
    return null
  },
}

export const validateVehicleData = (data) => {
  const errors = {}

  const makeErr = validators.textField(data.make || '', 'Make', 1)
  if (makeErr) errors.make = makeErr

  const modelErr = validators.textField(data.model || '', 'Model', 1)
  if (modelErr) errors.model = modelErr

  const engineSizeErr = validators.engineSize(data.engineSize)
  if (engineSizeErr) errors.engineSize = engineSizeErr

  const cylindersErr = validators.cylinders(data.cylinders)
  if (cylindersErr) errors.cylinders = cylindersErr

  const fuelCityErr = validators.fuelConsumption(data.fuelConsumptionCity, 'City fuel consumption')
  if (fuelCityErr) errors.fuelConsumptionCity = fuelCityErr

  const fuelHwyErr = validators.fuelConsumption(data.fuelConsumptionHighway, 'Highway fuel consumption')
  if (fuelHwyErr) errors.fuelConsumptionHighway = fuelHwyErr

  const fuelCombErr = validators.fuelConsumption(data.fuelConsumptionCombined, 'Combined fuel consumption')
  if (fuelCombErr) errors.fuelConsumptionCombined = fuelCombErr

  const classErr = validators.vehicleClass(data.vehicleClass)
  if (classErr) errors.vehicleClass = classErr

  const fuelErr = validators.fuelType(data.fuelType)
  if (fuelErr) errors.fuelType = fuelErr

  const transErr = validators.textField(data.transmission || '', 'Transmission', 1)
  if (transErr) errors.transmission = transErr

  return Object.keys(errors).length === 0 ? null : errors
}

File: backend/app/utils/validators.py
from typing import Tuple
from email_validator import validate_email, EmailNotValidError

class Validator:
    """Input validation utilities"""
    
    @staticmethod
    def validate_email(email: str) -> Tuple[bool, str]:
        """Validate email format"""
        try:
            valid = validate_email(email)
            return True, valid.email
        except EmailNotValidError as e:
            return False, str(e)
    
    @staticmethod
    def validate_registration(email: str, password: str, full_name: str) -> Tuple[bool, str]:
        """Validate registration data"""
        # Validate email
        is_valid, msg = Validator.validate_email(email)
        if not is_valid:
            return False, f"Invalid email: {msg}"
        
        # Validate password
        if len(password) < 8:
            return False, "Password must be at least 8 characters"
        if not any(c.isupper() for c in password):
            return False, "Password must contain uppercase letter"
        if not any(c.isdigit() for c in password):
            return False, "Password must contain digit"
        
        # Validate full name
        if len(full_name.strip()) < 2:
            return False, "Full name must be at least 2 characters"
        
        return True, "Valid"
    
    @staticmethod
    def validate_vehicle_data(
        make: str,
        model: str,
        engine_size: float,
        cylinders: int,
        fuel_consumption_city: float,
        fuel_consumption_highway: float,
        fuel_consumption_combined: float,
        fuel_type: str,
        vehicle_class: str,
        transmission: str
    ) -> Tuple[bool, str]:
        """Validate vehicle input data"""
        
        # Make and model have defaults, so just validate they're not empty
        if not make:
            make = "ACURA"
        if not model:
            model = "ILX"
        
        # Validate engine size
        if engine_size <= 0 or engine_size > 10:
            return False, "Engine size must be between 0.1 and 10 L"
        
        # Validate cylinders
        if cylinders < 1 or cylinders > 16:
            return False, "Cylinders must be between 1 and 16"
        
        # Validate fuel consumption
        if fuel_consumption_city <= 0 or fuel_consumption_city > 100:
            return False, "City fuel consumption must be between 0.1 and 100 L/100km"
        
        if fuel_consumption_highway <= 0 or fuel_consumption_highway > 100:
            return False, "Highway fuel consumption must be between 0.1 and 100 L/100km"
        
        if fuel_consumption_combined <= 0 or fuel_consumption_combined > 100:
            return False, "Combined fuel consumption must be between 0.1 and 100 L/100km"
        
        # Validate string fields
        valid_fuel_types = ["X", "Z", "D", "E", "N"]
        if fuel_type.upper() not in valid_fuel_types:
            return False, f"Fuel type must be one of: {', '.join(valid_fuel_types)}"
        
        valid_classes = [
            'COMPACT',
            'SUBCOMPACT',
            'MID-SIZE',
            'FULL-SIZE',
            'TWO-SEATER',
            'SUV - SMALL',
            'SUV - STANDARD',
            'PICKUP TRUCK - SMALL',
            'PICKUP TRUCK - STANDARD',
            'VAN - PASSENGER',
            'VAN - CARGO',
            'MINIVAN',
            'STATION WAGON - SMALL',
            'STATION WAGON - MID-SIZE',
        ]
        if vehicle_class.upper() not in valid_classes:
            return False, f"Vehicle class must be one of: {', '.join(valid_classes)}"
        
        if len(transmission.strip()) < 2:
            return False, "Transmission must be specified"
        
        return True, "Valid"
